src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/SamAuthProvider.scala

```diff
@@ -0,0 +1,218 @@
+package org.broadinstitute.dsde.workbench.leonardo.auth
+
+import akka.http.scaladsl.model.StatusCodes
+import com.google.api.client.googleapis.auth.oauth2.GoogleCredential
+import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport
+import com.google.api.client.json.jackson2.JacksonFactory
+import com.google.api.services.plus.PlusScopes
+import com.google.auth.oauth2.ServiceAccountCredentials
+import com.google.common.cache.{CacheBuilder, CacheLoader}
+import com.google.gson.internal.LinkedTreeMap
+import com.typesafe.config.Config
+import io.swagger.client.ApiClient
+import io.swagger.client.api.ResourcesApi
+import io.swagger.client.api.GoogleApi
+import java.io.{ByteArrayInputStream, File}
+import java.util.concurrent.TimeUnit
+import com.google.gson.Gson
+import com.typesafe.scalalogging.LazyLogging
+import org.broadinstitute.dsde.workbench.leonardo.model._
+import org.broadinstitute.dsde.workbench.leonardo.model.Actions._
+import org.broadinstitute.dsde.workbench.leonardo.model.NotebookClusterActions._
+import org.broadinstitute.dsde.workbench.leonardo.model.ProjectActions._
+import org.broadinstitute.dsde.workbench.model.WorkbenchEmail
+import net.ceedubs.ficus.Ficus._
+import scala.collection.JavaConverters._
+import scala.concurrent.{ExecutionContext, Future}
+
+case class UserEmailAndProject(userEmail: String, googleProject: String)
+
+case class NotebookActionError(action: Action) extends
+  LeoException(s"${action.toString} was not recognized", StatusCodes.NotFound)
+
+
+class SamAuthProvider(authConfig: Config, serviceAccountProvider: ServiceAccountProvider) extends LeoAuthProvider(authConfig, serviceAccountProvider) with LazyLogging {
+
+  private val notebookClusterResourceTypeName = "notebook-cluster"
+  private val billingProjectResourceTypeName = "billing-project"
+
+  private val httpTransport = GoogleNetHttpTransport.newTrustedTransport
+  private val jsonFactory = JacksonFactory.getDefaultInstance
+  private val saScopes = Seq(PlusScopes.USERINFO_EMAIL, PlusScopes.USERINFO_PROFILE)
+
+  //Cache lookup of pet tokens
+  val cacheExpiryTime = authConfig.getInt("cacheExpiryTime")
+  val cacheMaxSize = authConfig.getInt("cacheMaxSize")
+
+  private[leonardo] val petTokenCache = CacheBuilder.newBuilder()
+    .expireAfterWrite(cacheExpiryTime, TimeUnit.MINUTES)
+    .maximumSize(cacheMaxSize)
+    .build(
+      new CacheLoader[UserEmailAndProject, String] {
+        def load(userEmailAndProject: UserEmailAndProject) = {
+          getPetAccessTokenFromSam(userEmailAndProject.userEmail, userEmailAndProject.googleProject)
+        }
+      }
+    )
+
+  //Leo SA details -- needed to get pet keyfiles
+  private val (leoEmail, leoPem) : (WorkbenchEmail, File) = serviceAccountProvider.getLeoServiceAccountAndKey
+
+  //Given a pem, gets an access token
+  private def getAccessTokenUsingPem(email: WorkbenchEmail, pem: File): String = {
```

Comment:
And we should be able to remove the Google `Credential` stuff too, right?

Reply:
see above